osu.Game.Rulesets.Osu/Difficulty/Evaluators/SpeedEvaluator.cs

```diff
@@ -33,18 +34,28 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current)
             var osuNextObj = (OsuDifficultyHitObject?)current.Next(0);
 
             double strainTime = osuCurrObj.StrainTime;
-            double doubletapness = 1;
 
             // Nerf doubletappable doubles.
-            if (osuNextObj != null)
-            {
-                double currDeltaTime = Math.Max(1, osuCurrObj.DeltaTime);
-                double nextDeltaTime = Math.Max(1, osuNextObj.DeltaTime);
-                double deltaDifference = Math.Abs(nextDeltaTime - currDeltaTime);
-                double speedRatio = currDeltaTime / Math.Max(currDeltaTime, deltaDifference);
-                double windowRatio = Math.Pow(Math.Min(1, currDeltaTime / osuCurrObj.HitWindowGreat), 2);
-                doubletapness = Math.Pow(speedRatio, 1 - windowRatio);
-            }
+            double currDeltaTime = Math.Max(1, osuCurrObj.DeltaTime);
+
+            // It's easier to gallop if you have more time between doubles
+            // Get max between next and prev ratio to avoid nerfing triples
+            double speedRatio = Math.Max(getSpeedRatio(osuCurrObj, osuPrevObj), getSpeedRatio(osuCurrObj, osuNextObj));
+
+            // Can't doubletap if circles don't intersect
+            double normalizedDistance = Math.Min(1, osuCurrObj.LazyJumpDistance / (OsuDifficultyHitObject.NORMALISED_RADIUS * 2));
+            double distanceFactor = normalizedDistance < 0.5 ? 1.0 : 1 - Math.Pow((normalizedDistance - 0.5) / 0.5, 0.5);
+
+            // Use HitWindowGreat * 2, because even if you can't get 300 with doubletapping - you still can gallop
+            const double power = 2;
+            double windowRatio = Math.Pow(Math.Min(1, currDeltaTime / (osuCurrObj.HitWindowGreat * 2)), power);
+
+            // Nerf even more if you don't need to gallop anymore
+            double halfPoint = Math.Pow(0.5, power);
+            if (windowRatio < halfPoint)
+                windowRatio *= windowRatio / halfPoint;
```

Comment:
I don't really understand what's going on here

Reply:
2 part curve:
- first is for when you can't get 300 while perfectly doubletapping (hitwindow300 < deltaT < 2 * hitwindow300)
- second is for when can get 300 while perfectly doubletapping
I can add desmos for clarity